Use conda-forge bullet feedstock's float64/double precision libraries and provide Mac Arm support

KrisBecker opened the pull request:
## Description
This PR removes use of the usgs-astrogeology::bullet feedstock and adds use of the conda-forge bullet feedstock. This provides support for the [Mac Arm platform](#5758) and other platforms available from conda.

ISIS uses the double precision build of the Bullet library to provide high precision ray tracing. The USGS maintained a specific version of the Bullet library because it required a change to the bullet system to accommodate large numbers of facets/segment. This PR removes this requirement but does impose some restrictions. This [bullet commit](https://github.com/bulletphysics/bullet3/commit/fcef2c81f2478328fac9ce323fb3909adb83d234) made in 2023 contains the same modification made in the [USGS bullet version](https://github.com/DOI-USGS/ISIS3/blob/dev/isis/src/base/objs/BulletShapeModel/IsisBullet.h).

This change does enforce the restriction of 134,217,728 triangles per part of which there can be no more than 16 parts. For [NAIF DSKs](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/req/dsk.html#Appendix%20B%20---%20DSK%20Subsystem%20Limits), this means there can be **no** more than 16 segments per BDS file. Other improvements can be made but these changes preserve existing behavior and enforce the limits.

CMake support adopts explicit language to incorporate conda-forge `bullet >= 3.25` float64 double precision libraries. This requires some very specific assumptions made regarding how conda-forge bullet feedstock is implemented in the CMake package interface. It nearly works using @oleg-alexandrov's[suggestion](https://github.com/DOI-USGS/ISIS3/issues/5100#issuecomment-2784069547) of explicit MODULE mode but results in the correct compile definition but the wrong float version of the libraries. This PR provides the `BulletFloat64` target with a [FindBulletFloat64.cmake](https://github.com/DOI-USGS/ISIS3/compare/dev...UA-LPL:ISIS3:enhancement/conda-bullet-arm-fix?expand=1#diff-1f6945ad78ac549ec9a4fdafc11c7db6bc3da6516983182c89bbfc2d663aa878) file. Suggestions and alternatives are welcome.

## Related Issue
#5100
#5758

## How Has This Been Validated?
Note these tests were run on an Apple M3 Max Arm platform running Sonoma 14.7.2 with 32GB of memory, 14 cores, 30 GPUs.  The ISIS dev branch was cloned and code changes added in this PR. I used conda to install the required ISIS conda development system using Terminal in arm64 mode - not intel - with the command 
> (CondaBulletArmFix) arm64@maxm3 ISIS3 % conda env create -n CondaBulletArmFix -f environment_arm.yml

The branch was created from the dev branch on 2025-04-17 after the [Mac Arm Port](#5758) was merged. The ISISTESTDATA was updated but ISISDATA was not. All the ctests were run with the following results:

**(CondaBulletArmFix) arm64@maxm3 build % ctest -j8 --output-on-failure**
```
99% tests passed, 3 tests failed out of 2677

Label Time Summary:
apollo          =   3.17 sec*proc (4 tests)
app             = 1872.37 sec*proc (674 tests)
cassini         =   4.23 sec*proc (5 tests)
chandrayaan1    =   0.93 sec*proc (1 test)
clementine      =   4.11 sec*proc (4 tests)
dawn            =   1.15 sec*proc (2 tests)
galileo         =   0.59 sec*proc (1 test)
hayabusa        =   1.08 sec*proc (2 tests)
hayabusa2       =   0.53 sec*proc (1 test)
isis            = 245.12 sec*proc (314 tests)
juno            =   0.48 sec*proc (1 test)
kaguya          =   1.67 sec*proc (2 tests)
lo              =   1.21 sec*proc (3 tests)
lro             =   1.75 sec*proc (3 tests)
mariner         =   0.91 sec*proc (1 test)
messenger       =   0.35 sec*proc (1 test)
mex             =   1.88 sec*proc (2 tests)
mgs             =   2.77 sec*proc (4 tests)
mro             =   4.53 sec*proc (6 tests)
newhorizons     =   2.16 sec*proc (4 tests)
odyssey         =   1.31 sec*proc (2 tests)
rosetta         =   1.66 sec*proc (2 tests)
tgo             =   0.55 sec*proc (1 test)
unit            = 282.85 sec*proc (368 tests)
viking          =   0.36 sec*proc (1 test)
voyager         =   0.34 sec*proc (1 test)

Total Test time (real) = 471.29 sec

The following tests did not run:
	1113 - TempTestingFiles.FunctionalTestJitterfitDefault (Disabled)
	1156 - TempTestingFiles.UnitTestImageImporterTestJpeg (Disabled)
	1538 - DefaultCube.FunctionalTestNoprojExpand (Disabled)
	1872 - MroCtxCube.FunctionalTestCtxcalDefault (Disabled)
	1875 - MroCtxCube.FunctionalTestCtxcalIofFalse (Disabled)

The following tests FAILED:
	 87 - isis_unit_test_EmbreeTargetShape (Failed)
	1452 - DefaultCube.FunctionalTestCam2CamOffbody (Failed)
	2155 - Phocube.FunctionalTestPhocubeSunIlluminationMask (Failed)
Errors while running CTest
```

Output of the actual areas are included here:

**(CondaBulletArmFix) arm64@maxm3 build % ctest --rerun-failed  --output-on-failure**
```
Test project /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/build
    Start   87: isis_unit_test_EmbreeTargetShape
1/3 Test   #87: isis_unit_test_EmbreeTargetShape ...................***Failed    1.95 sec
-- ISISROOT = /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/build
-- ISISDATA = /opt/isis4/data
outputFile = /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/build/unitTest/isis_unit_test_EmbreeTargetShape.output
ln: /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/src/base/objs/EmbreeTargetShape/unitTest: File exists
-- Tolerance file: /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/src/base/objs/EmbreeTargetShape/EmbreeTargetShape.truth.TOL
-- Tolerance file does not exist. Using tolerance: 1e-8.
-- Comparison command: python /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/build/scripts/textdiff.py /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/build/unitTest/isis_unit_test_EmbreeTargetShape.output /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/src/base/objs/EmbreeTargetShape/EmbreeTargetShape.truth -abs_err 1e-8
------------------ COMPARISON OUTPUT ------------------ 
First file: /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/build/unitTest/isis_unit_test_EmbreeTargetShape.output
Second file: /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/src/base/objs/EmbreeTargetShape/EmbreeTargetShape.truth
ERROR: Got absolute error of 1.0000000000287557e-07 at line 205, word 3
First  file line: 205: Intersection:   ( 0.0497395   0.0997398   0 )
Second file line: 205: Intersection:   ( 0.0497396   0.0997398   0 )
ERROR: Got absolute error of 1.000000000001e-06 at line 209, word 3
First  file line: 209: Intersection:   ( -0.166992   -0.116992   0 )
Second file line: 209: Intersection:   ( -0.166993   -0.116992   3e-07 )
ERROR: Got absolute error of 3e-07 at line 209, word 5
First  file line: 209: Intersection:   ( -0.166992   -0.116992   0 )
Second file line: 209: Intersection:   ( -0.166993   -0.116992   3e-07 )
ERROR: Got absolute error of 1.2999999999999998e-06 at line 218, word 3
First  file line: 218: Intersection:   ( 1e-06   0.0579839   0.107985 )
Second file line: 218: Intersection:   ( -3e-07   0.0579836   0.107985 )
ERROR: Got absolute error of 2.999999999947489e-07 at line 218, word 4
First  file line: 218: Intersection:   ( 1e-06   0.0579839   0.107985 )
Second file line: 218: Intersection:   ( -3e-07   0.0579836   0.107985 )
ERROR: Got absolute error of 1e-07 at line 222, word 3
First  file line: 222: Intersection:   ( 1e-07   -0.111983   -0.0619813 )
Second file line: 222: Intersection:   ( 0   -0.111983   -0.0619813 )
ERROR: Got absolute error of 3.000000000016878e-07 at line 235, word 3
First  file line: 235: Intersection:   ( -0.0527439   1e-07   -0.102744 )
Second file line: 235: Intersection:   ( -0.0527436   3e-07   -0.102744 )
ERROR: Got absolute error of 2e-07 at line 235, word 4
First  file line: 235: Intersection:   ( -0.0527439   1e-07   -0.102744 )
Second file line: 235: Intersection:   ( -0.0527436   3e-07   -0.102744 )
ERROR: Got absolute error of 5.000000000005e-07 at line 244, word 3
First  file line: 244: Intersection:   ( 0.0831601   0.0831605   0.0831612 )
Second file line: 244: Intersection:   ( 0.0831606   0.0831606   0.083161 )
ERROR: Got absolute error of 1.0000000000287557e-07 at line 244, word 4
First  file line: 244: Intersection:   ( 0.0831601   0.0831605   0.0831612 )
Second file line: 244: Intersection:   ( 0.0831606   0.0831606   0.083161 )
ERROR: Got absolute error of 2.0000000000575113e-07 at line 244, word 5
First  file line: 244: Intersection:   ( 0.0831601   0.0831605   0.0831612 )
Second file line: 244: Intersection:   ( 0.0831606   0.0831606   0.083161 )
ERROR: Got absolute error of 9.999999998899778e-08 at line 248, word 4
First  file line: 248: Intersection:   ( -0.0834811   -0.0834812   -0.0834812 )
Second file line: 248: Intersection:   ( -0.0834811   -0.0834813   -0.0834812 )
Input tolerance: 1e-08
Max observed absolute error: 1.2999999999999998e-06
------------------------------------------------- 
CMake Error at /Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/cmake/RunUnitTest.cmake:85 (message):
  Test failed.  Files differ with tolerance.



    Start 1452: DefaultCube.FunctionalTestCam2CamOffbody
2/3 Test #1452: DefaultCube.FunctionalTestCam2CamOffbody ...........***Failed    0.42 sec
Note: Google Test filter = DefaultCube.FunctionalTestCam2CamOffbody
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DefaultCube
[ RUN      ] DefaultCube.FunctionalTestCam2CamOffbody
/Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/tests/FunctionalTestsCam2cam.cpp:346: Failure
Value of: map_from_default.mapit( v_from_sample, v_from_line, v_match_sample, v_match_line)
  Actual: true
Expected: false

/Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/tests/FunctionalTestsCam2cam.cpp:347: Failure
Value of: map_from_off_notrim.mapit( v_from_sample, v_from_line, v_match_sample, v_match_line)
  Actual: true
Expected: false

/Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/tests/FunctionalTestsCam2cam.cpp:348: Failure
Value of: map_from_off_trim.mapit( v_from_sample, v_from_line, v_match_sample, v_match_line)
  Actual: true
Expected: false

[  FAILED  ] DefaultCube.FunctionalTestCam2CamOffbody (215 ms)
[----------] 1 test from DefaultCube (215 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (215 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DefaultCube.FunctionalTestCam2CamOffbody

 1 FAILED TEST

    Start 2155: Phocube.FunctionalTestPhocubeSunIlluminationMask
3/3 Test #2155: Phocube.FunctionalTestPhocubeSunIlluminationMask ...***Failed    1.00 sec
Note: Google Test filter = Phocube.FunctionalTestPhocubeSunIlluminationMask
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from Phocube
[ RUN      ] Phocube.FunctionalTestPhocubeSunIlluminationMask
/Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/tests/FunctionalTestsPhocube.cpp:502: Failure
The difference between hist->Average() and 0.159 is 0.083400000000000002, which exceeds .000001, where
hist->Average() evaluates to 0.075600000000000001,
0.159 evaluates to 0.159, and
.000001 evaluates to 9.9999999999999995e-07.

/Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/tests/FunctionalTestsPhocube.cpp:503: Failure
The difference between hist->Sum() and 1590 is 834, which exceeds .000001, where
hist->Sum() evaluates to 756,
1590 evaluates to 1590, and
.000001 evaluates to 9.9999999999999995e-07.

/Users/kbecker/ISIS/GitHubCheckouts/CondaBulletArmFix/ISIS3/isis/tests/FunctionalTestsPhocube.cpp:505: Failure
The difference between hist->StandardDeviation() and 0.3656943713503446 is 0.10132412120310075, which exceeds .000001, where
hist->StandardDeviation() evaluates to 0.26437025014724386,
0.3656943713503446 evaluates to 0.3656943713503446, and
.000001 evaluates to 9.9999999999999995e-07.

[  FAILED  ] Phocube.FunctionalTestPhocubeSunIlluminationMask (888 ms)
[----------] 1 test from Phocube (888 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (888 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] Phocube.FunctionalTestPhocubeSunIlluminationMask

 1 FAILED TEST


0% tests passed, 3 tests failed out of 3

Label Time Summary:
isis    =   1.95 sec*proc (1 test)
unit    =   1.95 sec*proc (1 test)

Total Test time (real) =   3.44 sec

The following tests FAILED:
	 87 - isis_unit_test_EmbreeTargetShape (Failed)
	1452 - DefaultCube.FunctionalTestCam2CamOffbody (Failed)
	2155 - Phocube.FunctionalTestPhocubeSunIlluminationMask (Failed)
Errors while running CTest
```
Not sure what to make of these errors. The `DefaultCube.FunctionalTestCam2CamOffbody` is likely related to the bullet changes.

### Residual Test Data Files
After ISIS is built there is what appears to be temporary files created from isis tests. Here is a summary of the files in this PR which shows the residual untracked files in the test directory.

![Screenshot 2025-04-17 at 4 09 24 PM](https://github.com/user-attachments/assets/753e828d-5d6b-4e17-ab16-ad69cf2eac13)


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Infrastructure change (changes to things like CI or the build system that do not impact users)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ ] My code follows the code style of this project. -->
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added my user impacting change to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document. <!--- NOTE: You can only have one changelog entry per PR, see https://github.com/DOI-USGS/ISIS3/blob/dev/CONTRIBUTING.md -->


## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

KrisBecker committed:
Update to use conda-forge bullet configuration

KrisBecker committed:
CMake configuration for BulletFloat64 target

KrisBecker committed:
Modifications to use bullet >= 3.25

oleg-alexandrov commented:
Kris, thanks.

There are fewer test failures than I expected, but the ones that fail seem to do so by large-ish values. 

I am not sure what to make of the failures either. Not sure if they are fully relevant for this package.

Does the updated ISIS work well enough for your regular use?

I will be back to ISIS work in a couple of weeks and will study this more, unless looked at before that.

KrisBecker committed:
Merge remote-tracking branch 'origin/dev' into enhancement/conda-bullet-arm-fix

KrisBecker committed:
Merge branch 'dev' of https://github.com/UA-LPL/ISIS3 into enhancement/conda-bullet-arm-fix

acpaquette commented:
@KrisBecker Did some digging on the tests, as we talked about briefly the last bullet release on conda was built with `#define MAX_NUM_PARTS_IN_BITS 10` which results in the bullet model not being able to load the DSK. For both of the tests this results in using a `NaifDskShape` which changes the results of the test. Below is a link to the include file that defines `MAX_NUM_PARTS_IN_BITS` for the latest 3.25 bullet release.

https://github.com/bulletphysics/bullet3/blob/3.25/src/BulletCollision/BroadphaseCollision/btQuantizedBvh.h

We can either poke the bullet maintainers to draft a new release and get a 3.26 build on conda-forge or update the bullet reader to create multiple parts from one DSK segment.

KrisBecker committed:
Merge remote-tracking branch 'origin/dev' into enhancement/conda-bullet-arm-fix

oleg-alexandrov commented:
@Kelvinrr : From what @acpaquette is saying, the conda-forge bullet won't work with ISIS no matter what as it is not built properly.

It is then puzzling that it has worked for @KrisBecker in his own work, but then the tests fail so badly that it appears bullet cannot even load the shape.

This really needs input from people who actually employ this logic and are knowledgeable enough about the particulars.

KrisBecker commented:
I am working on a solution that alters how the mesh data is mapped into the Bullet system. This change must consider limits of any Bullet configuration of the part size and data volume. I have retested my version and it does need this change in order to fix the limitations of the current configuration in bullet>=3.25. It impacts all OS versions.

I expect it to be done by the end of the week.

KrisBecker committed:
Fix mapping of DSK segments to Bullet mesh reqs

- DskSegmentBuffer.hpp - Added new class  which manages the facet data read from the DSK file for each segment and creates Bullet indexed meshes from the data
- BulletDskShape.h,.cpp - Updated for new DSK data management implementation; reworked how the DSK data is managed; checks for Bullet mesh array (part) size exceeding limits for quantization/optimizations - In this cases, turn off this optimization and issue a warning. No errors are generated.
- BulletTargetShape.h,.cpp - Updated to support use of c++17 STL for memory management rather than Qt; added cleanup of Bullet collision object.

KrisBecker committed:
Fix failing EnbreeTargetShape unit test

- EmbreeTargetShape.truth.TOL - Added tolerance file for unittests so they pass on the Mac platform. The new tolerance is 1.3e-6. This kinda makes sense because Embree only supports floating point data as welll as ray tracing internal implementations in floating point.

KrisBecker committed:
Update conda YAML config files for consistency

KrisBecker committed:
Merge branch 'dev' into enhancement/conda-bullet-arm-fix

KrisBecker committed:
Add CHANGELOG entry for #5772

KrisBecker commented:
Ok, this is ready for another review. Here is a rundown of the main changes made:

#### Fix mapping of DSK segments to Bullet mesh reqs
- DskSegmentBuffer.hpp - Added new class  which manages the facet data read from the DSK file for each segment and creates Bullet indexed meshes from the data
- BulletDskShape.h,.cpp - Updated for new DSK data management implementation; reworked how the DSK data is managed; checks for Bullet mesh array (part) size exceeding limits for quantization/optimizations - In these cases, turns off this optimization and issues a warning. No errors are generated. Should work for any Bullet configuration.
- BulletTargetShape.h,.cpp - Updated to support use of c++17 STL for memory management rather than Qt; added memory (leak) cleanup of Bullet collision object.

#### Fix failing EnbreeTargetShape unit test
- EmbreeTargetShape.truth.TOL - Added tolerance file for unittests so they pass on the Mac platform. The new tolerance is 1.3e-6. This kinda makes sense because Embree only supports floating point data as welll as ray tracing internal implementations in floating point.

All tests are currently passing on the Mac Arm, so I made a change log entry. Its ready to merge.

oleg-alexandrov commented on isis/src/base/objs/BulletDskShape/BulletDskShape.cpp:
```diff
@@ -155,14 +145,29 @@ namespace Isis {
+ * fit into the bullet mesh mapping scheme which did have 1024 parts that supports no more
+ * than 2,097,152 triangles. At ant rate, loading DSKs must consider these limits when mapping to
+ * bullet use. So DSK support is limited to at most 1024 segments where each segment has no more
+ * than 2,097,152. Since 
```
This comment likely needs a bit of work when it comes to: "ant rate", "2,097,152.", and "Since.", so fixing one typo, completing one sentence, and removing unneeded word at the end.

oleg-alexandrov reviewed (approved):
Kris, thanks. Looks good to me.

KrisBecker committed:
Fix typos and clarify documentation

Kelvinrr reviewed (approved):
looks good to me with comment addressed. Thanks Kris for taking care of this, it's a huge benefit to ISIS.
